From e2c0ea4374c02539ad9bc6c4328b74f25eb72405 Mon Sep 17 00:00:00 2001
From: The Datalorax <datalorax@gmail.com>
Date: Tue, 28 Apr 2015 21:07:21 +0100
Subject: [PATCH 1/2] #LAN-1114 fixes bug in TypeUtils.equals(WildcardType,
 Type) where it was incorrectly returning true when the second argument was
 not a Wildcard type.

---
 .../org/apache/commons/lang3/reflect/TypeUtils.java   |  2 +-
 .../apache/commons/lang3/reflect/TypeUtilsTest.java   | 11 +++++++++++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java b/src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java
index 1a750e97dfb..91313c6a473 100644
--- a/src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java
+++ b/src/main/java/org/apache/commons/lang3/reflect/TypeUtils.java
@@ -1626,7 +1626,7 @@ private static boolean equals(final WildcardType w, final Type t) {
             return equals(getImplicitLowerBounds(w), getImplicitLowerBounds(other))
                 && equals(getImplicitUpperBounds(w), getImplicitUpperBounds(other));
         }
-        return true;
+        return false;
     }
 
     /**
diff --git a/src/test/java/org/apache/commons/lang3/reflect/TypeUtilsTest.java b/src/test/java/org/apache/commons/lang3/reflect/TypeUtilsTest.java
index 60c8b4897b5..56b23a2411a 100644
--- a/src/test/java/org/apache/commons/lang3/reflect/TypeUtilsTest.java
+++ b/src/test/java/org/apache/commons/lang3/reflect/TypeUtilsTest.java
@@ -97,6 +97,8 @@ public class Tester implements This<String, B> {
 
     public static Comparable<Long> longComparable;
 
+    public static Comparable<?> wildcardComparable;
+
     public static URI uri;
 
     public void dummyMethod(final List list0, final List<Object> list1, final List<?> list2,
@@ -722,6 +724,15 @@ public void testLowerBoundedWildcardType() {
        Assert.assertEquals(String.format("? super %s", iterableT0.getName()), lowerTypeVariable.toString());
     }
 
+    @Test
+    public void testLang1114() throws Exception {
+        final Type nonWildcardType = getClass().getDeclaredField("wildcardComparable").getGenericType();
+        final Type wildcardType = ((ParameterizedType)nonWildcardType).getActualTypeArguments()[0];
+
+        Assert.assertFalse(TypeUtils.equals(wildcardType, nonWildcardType));
+        Assert.assertFalse(TypeUtils.equals(nonWildcardType, wildcardType));
+    }
+
     @Test
     public void testGenericArrayType() throws Exception {
         final Type expected = getClass().getField("intWildcardComparable").getGenericType();

From 6271c5015feecc22b9197ef07170fa15f311aad1 Mon Sep 17 00:00:00 2001
From: Benedikt Ritter <britter@apache.org>
Date: Wed, 29 Apr 2015 19:45:35 +0200
Subject: [PATCH 2/2] Add LANG-1114 to changes.xml

---
 src/changes/changes.xml | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 3e25e069451..1a59e4658f9 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -22,6 +22,7 @@
   <body>
 
   <release version="3.5" date="tba" description="tba">
+    <action issue="LANG-1114" type="fix" dev="britter" due-to="Andy Coates">TypeUtils.ParameterizedType#equals doesn't work with wildcard types</action>
     <action issue="LANG-979" type="update" dev="britter" due-to="Bruno P. Kinoshita">TypeUtils.parameterizeWithOwner - wrong format descriptor for "invalid number of type parameters".</action>
     <action issue="LANG-1112" type="update" dev="britter">MultilineRecursiveToStringStyle largely unusable due to being package-private.</action>
     <action issue="LANG-1058" type="update" dev="djones" due-to="Leo Wang">StringUtils.uncapitalize performance improvement</action>
